src/node_url.cc

```diff
@@ -407,29 +407,86 @@ int64_t ParseIPv4Number(const char* start, const char* end) {
   return strtoll(start, nullptr, R);
 }
 
+// https://url.spec.whatwg.org/#ipv4-number-parser
+bool IsIPv4NumberValid(std::string_view input) {
+  // If input is the empty string, then return failure.
+  if (input.empty()) {
+    return false;
+  }
+
+  // If input contains at least two code points..
+  if (input.size() >= 2) {
+    // and the first two code points are either "0X" or "0x", then:
+    if (input[0] == '0' && (input[1] == 'X' || input[1] == 'x')) {
+      if (input.size() == 2) {
+        return true;
+      }
+
+      // Remove the first two code points from input,
+      // radix-R is 16
+      // If input contains a code point that is not a radix-R digit, then return
+      // failure.
+      return input.find_first_not_of("0123456789abcdefABCDEF", 2) ==
+             std::string_view::npos;
+
+      // and the first code point is U+0030 (0), then:
+    } else if (input[0] == '0') {
+      if (input.size() == 1) {
+        return true;
+      }
+
+      // Remove the first code point from input.
+      // radix-R is 8
+      // If input contains a code point that is not a radix-R digit, then return
+      // failure.
+      return input.find_first_not_of("01234567", 1) == std::string_view::npos;
+    }
+  }
+
+  // If input contains a code point that is not a radix-R digit, then return
+  // failure. radix-R is 10
+  return std::all_of(input.begin(), input.end(), ::isdigit);
+}
+
 // https://url.spec.whatwg.org/#ends-in-a-number-checker
-bool EndsInANumber(const std::string& input) {
-  std::vector<std::string> parts = SplitString(input, '.', false);
+bool EndsInANumber(const std::string_view input) {
+  if (input.empty()) {
+    return false;
+  }
+
+  const std::string delimiter = ".";
```

Comment:
we can just use a `char` in here.

Reply:
actually i think this needs to be a string so i can use `std::find_end`

Reply:
but I'll confirm this

Reply:
You can use `string_view::rfind` with char